### PR TITLE
Base zoom speed on browser refresh rate

### PR DESCRIPTION
**Pull request by lukemckinstry**
# Description

Bases zoom speed on the render rate of the user's browser.
Added fields in FrameRateMonitor to track the timestamps of the n most recent renders.
Then when calculating the zoom distance on each frame, adjust it based on the fps zoom speed to provide consistent experience for browsers with different speeds.
If the user wants faster or slower zoom, they can still adjust zoomFactor through the public api as before with the same effects.

## Issue number and link
https://github.com/CesiumGS/cesium/issues/12187

## Testing plan

# Author checklist

- [x] I have submitted a Contributor License Agreement
- [x] I have added my name to `CONTRIBUTORS.md`
- [x] I have updated `CHANGES.md` with a short summary of my change
- [x] I have added or updated unit tests to ensure consistent code coverage
- [x] I have updated the inline documentation, and included code examples where relevant
- [x] I have performed a self-review of my code

### CODE REVIEW

**Comment by lukemckinstry** [comment]
posted this for initial review/feedback/discussion @ggetz. I still need to add unit tests.

**Comment by javagl** [comment]
This looks somewhat similar to what I dumped into [a forum thread](https://community.cesium.com/t/fps-drop-while-rendering-sparse-point-cloud-data-as-3d-tiles/28657/8) a while ago. The goal _there_ was also to have an "average frame rate over the last `n` frames". 

In that forum thread, I put this into some `FpsTracker` class. The reasons may be obvious. The functionality is _encapsulated_ in a class where the name says what the class does (`/* An FpsTracker tracks the FPS */` 😁  ). It offers a `getAverageFps` function _directly_. It avoids that `const timeStamps = object._scene.frameState?.timeStamps;` (aka [Train Wreck](https://wiki.c2.com/?TrainWreck)). And it avoids the manual computation (with magic constants like `10` and `30`) inside the `handleZoom` function.

The purpose of that `FpsTracker` there was to plug it into some `MsseUpdater`, which was supposed the update the maximum screen space error based on the average FPS. (Fleshing this out into a _real feature_ is still on my TODO list - I think that this could make a whole lot of sense...). 

**tl;dr** : If there was "something" where a function like `getAverageFps` was offered, then this could probably be re-used in multiple places...

**Comment by lukemckinstry** [comment]
I like your idea to create a `FpsTracker` class to make this more reusable. I will work on this change and update this PR when it is ready.

**Comment by ggetz** [comment]
I like the idea of isolating the average FPS logic in one place. The update itself should probably still happen in `Scene.js` though.

File: packages/engine/Source/Scene/Scene.js [hunk 1]
```diff
@@ -162,6 +163,7 @@ function Scene(options) {
   this._frameState.scene3DOnly = defaultValue(options.scene3DOnly, false);
   this._removeCreditContainer = !hasCreditContainer;
   this._creditContainer = creditContainer;
+  this._fpsTracker = new FpsTracker();
```
Comment (ggetz) [inline packages/engine/Source/Scene/Scene.js:166]: Since we are accessing this from another class, can we expose this as `fpsTracker` with a getter? It can be marked as `@private` so it doesn't show up in the developer API documentation.

Comment (lukemckinstry) [reply packages/engine/Source/Scene/Scene.js:166]: updated with this change :+1:

File: packages/engine/Source/Core/FpsTracker.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ * @private
+ */
+function FpsTracker(options) {
+  this._timeStamps = [];
+}
+
+/**
+ * Calculates average fps based on timeStamps of recently rendered frames
+ */
+FpsTracker.prototype.getAverageFps = function () {
```
Comment (javagl) [inline packages/engine/Source/Core/FpsTracker.js:14]: The hard-coded `10` (and the corresponding `9`s below) make my eye twitch 🙁 

But beyond that: This does not really compute the "average FPS". 

Consider time stamps like
`1 2 3 4 5 6 7 8 9 10`
It will compute `(10 - 1) / 9 = 1.0`. 
This means that the ~"time per frame" is `1.0`. (This isn't really the "elapsed time", but ... well...)

For
`1 2 3 4 5 6 7 8 9 19`
it will compute `(19 - 1) / 9 = 2.0`.
So even though the "time per frame" will constantly be `1.0` (except for the outlier of the last frame), the "time per frame" will be computed as `2.0`. The whole idea behind storing multiple time stamps is _exactly_ to "smooth (average) out" single outliers like that last `19`. 

One could mention that, in the current approach, it wouldn't even be necessary to store the 10 time stamps. Storing **2** (and a 'counter') would be enough. For a proper "sliding window", one could examine the performance of `unsihift`/`pop` vs. `push`/`shift`. Or one avoid these considerations altogether and use a fixed-sized array and a 'cursor' (aka "circular buffer").

File: packages/engine/Source/Core/FpsTracker.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+FpsTracker.prototype.updateFrameTimeStamp = function () {
+  const time = getTimestamp();
+  this._timeStamps.unshift(time);
+  if (this._timeStamps.length > 10) {
+    this._timeStamps.pop();
+  }
+};
+
+/**
+ * Destroys the WebGL resources held by this object.
```
Comment (javagl) [inline packages/engine/Source/Core/FpsTracker.js:35]: This does not have WebGL resources

Comment (lukemckinstry) [reply packages/engine/Source/Core/FpsTracker.js]: thank you, since we will now use the FrameRateMonitor class this is no longer relevant

File: packages/engine/Source/Core/FpsTracker.js (hunk 2 above)
Comment (javagl) [reply packages/engine/Source/Core/FpsTracker.js]: This should be obsolete with the `FrameRateMonitor`. 

(While some aspects of this class are hard to understand at a glimpse (like that "visibility" thing....), it seems to offer the required functionality, so _not_ duplicating that (even in a simpler, more self-contained form) seems reasonable)

**Comment by lukemckinstry** [comment]
We realized the existing FrameRateMonitor class provides close to the functionality this feature needs. We decided to make some small additions to it rather than create a new class (side notes: All the work up to this commit 331650c67d80b9011ec15a077fa7c8ea6e8b6f6d got re-written, so it may make sense to squash the early commits )

There may be questions as to why we dont get FPS from `FrameRateMonitor.lastFramesPerSecond`, this should be made clear by the doc updates included in this PR.

One open comment that I have not addressed: Including the target/idea browser refresh rate as constant value (30 as in 30hz) in the code. I am unsure if we should A) expose setting this in the API (but we already have zoomFactor in the public API) or B) set it in a more clear place, like a constants or system settings file.

File: packages/engine/Source/Scene/ScreenSpaceCameraController.js [hunk 4]
```diff
@@ -573,7 +576,16 @@ function handleZoom(
   const maxHeight = object.maximumZoomDistance;
 
   const minDistance = distanceMeasure - minHeight;
-  let zoomRate = zoomFactor * minDistance;
+
+  let fpsMultiplier = 1;
+  const fps = object.frameRateMonitor.averageFramesPerSecond;
+  // we set the ideal browser refresh rate to 30hz
+  if (fps) {
+    fpsMultiplier = 30 / fps;
```
Comment (lukemckinstry) [inline packages/engine/Source/Scene/ScreenSpaceCameraController.js:584]: This is one open comment that I have not addressed: Including the target/ideal browser refresh rate as constant value (30 as in 30hz) in the code here. I am unsure if we should A) expose setting this in the API (but we already have zoomFactor in the public API) or B) set it in a more clear place, like a constants or system settings file.
I prefer option B

**Review by ggetz — COMMENTED** [review]
Looking good @lukemckinstry! I have a few comments on code style, and a few thoughts on behavior.

**Comment by ggetz** [comment]
Please excuse the comments on outdated code, I'm not sure how they snuck in there.

File: packages/engine/Source/Scene/ScreenSpaceCameraController.js (hunk 4 above)
Comment (lukemckinstry) [reply packages/engine/Source/Scene/ScreenSpaceCameraController.js]: I added a check to make sure `fps > 0`, which should ensure we avoid division by 0

File: packages/engine/Source/Scene/ScreenSpaceCameraController.js [hunk 5]
```diff
@@ -573,7 +576,16 @@ function handleZoom(
...
+  }
+
+  let zoomRate = zoomFactor * minDistance * fpsMultiplier;
```
[the remark this answers is not in the document]
Comment (ggetz) [reply packages/engine/Source/Scene/ScreenSpaceCameraController.js:587]: I like this solution. I wonder if we can set the min to something like `0.5` to still account for time when the FPS does dip a bit.

[the remark this answers is not in the document]
Comment (javagl) [reply packages/engine/Source/Scene/ScreenSpaceCameraController.js:587]: I might be lacking some context, but would have thought that taking the _average_ FPS was supposed to smooth out any hiccups in the FPS.

[the remark this answers is not in the document]
Comment (lukemckinstry) [reply packages/engine/Source/Scene/ScreenSpaceCameraController.js:587]: Happy to fill in context @javagl 
The situation we are seeing when we test the branch on scenes with heavy data loading operations (like this sandcastle using the NY Buildings data https://sandcastle.cesium.com/?src=3D%20Tiles%20Feature%20Picking.html) is these heavy operations can make the render rate can slow down for a while (several seconds) or even indefinitely if the user continues operations that forces data reloading.

Low FPS leads to a high fpsMulitplier. 

In the equation `let zoomRate = zoomFactor * minDistance * fpsMultiplier` minDistance is supposed to be small as we get close to the ellipsoid surface and large when far out in space to create intuitive UX. But high fpsMultiplier throws this off, yielding a high zoomRate when we are already zoomed in close, which explains the jerky/jumpy behavior we were referring to.

Back to the original complaint/issue. The problem is actually not when FPS is low, it is only when FPS is too fast. So the idea with `fpsMultiplier = Math.min(fpsMultiplier, 1.0)` is to only adjust the zoomRate downward towards what we expect from a 60 FPS screen when the screen has a much higher refresh rate (eg. 120hz).

**Comment by ggetz** [comment]
@lukemckinstry I'm going to close this to keep things tidy. If you or anyone else wants to update and finish this out, please feel free to re-open!
